src/flint/types/fmpz.pyx

```diff
@@ -683,11 +683,30 @@ cdef class fmpz(flint_scalar):
         return fmpz_is_probabprime(self.val)
 
     def is_perfect_power(self):
+        r"""
+        Return True if this integer is of the form `r^k`, False otherwise.
```

Comment:
We need to clarify what the convention is here. At least we need `k > 1` for this to be meaningful.

It seems to me that it is only really useful to say that a number is a perfect power if `r != self` and `k > 1` although that is not how flint's `is_perfect_power` function handles -1, 0 and 1.

Reply:
I agree that `k>1` may have to be added; it is also missing in the C library documentation, but I doubt that anyone has been confused because of it. Regarding -1, 0, 1: that is explicitly mentioned in the documentation, https://flintlib.org/doc/fmpz.html
I seems odd to deviate from the C library behavior in Python bindings.

Reply:
> I seems odd to deviate from the C library behavior in Python bindings.

I agree. Also the `is_perfect_power` method has been there a long time so it should not be changed without a strong reason.

I just have a hard time imagining that defining 0 and 1 to be perfect powers is particularly useful for anything. In practice the value of knowing that something is a perfect power is if it means that you can reduce a problem down to a smaller number. Finding perfect powers of negative numbers is also not really that useful because it is better to handle the sign separately and find a power representation for the absolute value:
```python
In [24]: perfect_power(-4)
Out[24]: False

In [25]: perfect_power(4)
Out[25]: (2, 2)
```

Reply:
Regarding tests, `fmpz_is_perfect_power` appears well-tested already: https://github.com/flintlib/flint/blob/main/src/fmpz/test/t-is_perfect_power.c

Reply:
The testing needed in python-flint is a bit different from the testing needed in Flint. Here in python-flint we can usually assume that what Flint does is mathematically correct so we don't need to exhaustively test different values.

Firstly we need at least one test just to see that python-flint does not segfault because of a bug in its own wrapper code.

Secondly though the problem is that Flint does things like segfaulting or aborting as a way of handling invalid inputs. For example it would not have surprised me if `fmpz(-8).is_perfect_power()` would just crash in much the same way that this does:
```python
In [28]: flint.nmod_poly([1, 2, 1], 4).factor()
Flint exception (Impossible inverse):
    Cannot invert modulo 2*2
Aborted (core dumped)
```
We generally want to avoid exposing that core dump to Python users because it is not friendly if you are working e.g. in a notebook or something. That means that python-flint needs a bit more validation of arguments to raise exceptions before calling Flint. It also means that python-flint needs to test edge cases to make sure that Flint behaves reasonably for the values that are passed to Flint.

Lastly there can often be reasons to have different behaviour in python-flint as compared to Flint because the most natural/convenient way of handling edge cases might be different in Python.

Reply:
There is such a test for `is_perfect_power` [here](https://github.com/flintlib/python-flint/blob/591f900c2b8824c4e950a39afca33321b1c36192/src/flint/test/test_all.py#L295C34-L295C50)

Do you mean this should be expanded further?
Added test for `is_square`

Reply:
> Do you mean this should be expanded further?

No, I just forgot that `is_perfect_power` was already there.